### PR TITLE
Feat/add repository for invite service

### PR DESCRIPTION
**Pull request by skelet0no**
1. Added repository files
2. Change docstrings
3. Added methods in invite.py
4. Added db connecting python file
5. Added value not found file in errors
6. Added prisma schema
7. Added class decorator singleton

### CODE REVIEW

File: invite_service/repository/mock_invite_repository.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        List of invites
+
+    Methods
+    -------
+    async get_invites(author_id)
+        Returns invites that has matches with given author id.
+    async create_invite(invite)
+        Creates new invite inside db.
+    async update_invite(invite)
+        Updates invite that has the same id as provided invite object inside db or throws an exception.
```
Comment (ZixMai) [inline invite_service/repository/mock_invite_repository.py]: remove "or throws an exception" in both methods

File: invite_service/repository/mock_invite_repository.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+            Author's id.
+
+        Returns
+        -------
+        List[Invite]
+            List of invites that matches by author id.
+
+        Raises
+        ------
+        prisma.errors.PrismaError
```
Comment (ZixMai) [inline invite_service/repository/mock_invite_repository.py]: this method doesnt raise any error from prisma remove it pls

File: invite_service/repository/mock_invite_repository.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        Updates invite with matching id or throws an exception
+
+        Parameters
+        ----------
+        invite : Invite
+            Invite data
+
+        Raises
+        ------
+        Value Not Found Error
```
Comment (ZixMai) [inline invite_service/repository/mock_invite_repository.py]: Error name must be written without empty spaces

File: invite_service/repository/mock_invite_repository.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        Deletes invite with matching id or throws an exception
+
+        Parameters
+        ----------
+        invite_id : str
+            Invite's id
+
+        Raises
+        ------
+        Value Not Found Error
```
Comment (ZixMai) [inline invite_service/repository/mock_invite_repository.py]: Error name must be written without empty spaces

File: invite_service/repository/invite_repository_impl.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+        db_invites: Optional[
+            List[PrismaInvite]
+        ] = await self._db_client.db.invite.find_many(where={"id": author_id})
+        if db_invites is None or len(db_invites) == 0:
+            raise ValueNotFoundError("Invites not found")
+        return [
+            Invite.from_prisma_invite(prisma_invite=db_invite) for db_invite in db_invites
+        ]
+
+    async def get_all_invites(self, invite: Invite) -> List[Invite]:
```
Comment (ZixMai) [inline invite_service/repository/invite_repository_impl.py]: you dont need invite to request all invites, remove it from parameters and from docstring

File: invite_service/src/invite_service_impl.py
Comment (ZixMai) [inline invite_service/src/invite_service_impl.py:1]: Add get all invites method to invite service impl file and into class docstring and add repository to init

File: invite_service/repository/mock_invite_repository.py
Comment (ZixMai) [inline invite_service/repository/mock_invite_repository.py:1]: Add get all invites and get invites by invitee id methods to file and into class docstring

File: invite_service/repository/invite_repository_interface.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        prisma.errors.PrismaError
+            Catch all for every exception raised by Prisma Client Python.
+        ValueNotFoundError
+            No invites was found for given author id.
+
+        """
+        pass
+
+    @abstractmethod
+    async def get_all_invites(self, invite: Invite) -> List[Invite]:
```
Comment (ZixMai) [inline invite_service/repository/invite_repository_interface.py]: we dont need invite object to request all invites, remove it from parameters

File: invite_service/repository/invite_repository_impl.py [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+    -------
+    async get_invites(author_id)
+        Returns invites that has matches with given author id.
+    async create_invite(invite)
+        Creates new invite inside db or throws an exception.
+    async update_invite(incite)
+        Updates invite that has the same id as provided invite object inside db or throws an exception.
+    async delete_invite(invite_id)
+        Deletes invite that has matching id from database or throws an exception.
+    async get_all_invites(invite)
```
Comment (ZixMai) [inline invite_service/repository/invite_repository_impl.py]: remove invite from parameters

File: invite_service/repository/invite_repository_interface.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+    -------
+    async get_invites(author_id)
+        Returns invites that has matches with given author id.
+    async create_invite(invite)
+        Creates new invite inside db or throws an exception.
+    async update_invite(invite)
+        Updates invite that has the same id as provided invite object inside db or throws an exception.
+    async delete_invite(invite_id)
+        Deletes invite that has matching id from database or throws an exception.
+    async get_all_invites(invite)
```
Comment (ZixMai) [inline invite_service/repository/invite_repository_interface.py]: we dont need invite object to request all invites, remove it from parameters

File: invite_service/repository/invite_repository_impl.py [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+        ------
+        prisma.errors.PrismaError
+            Catch all for every exception raised by Prisma Client Python.
+        ValueNotFoundError
+            No invites was found for given author id.
+
+        """
+        db_invites: Optional[
+            List[PrismaInvite]
+        ] = await self._db_client.db.invite.find_many(where={"id": author_id})
```
Comment (ZixMai) [inline invite_service/repository/invite_repository_impl.py]: check if deleted_at is None

File: invite_service/src/models/invite.py
Comment (ZixMai) [inline invite_service/src/models/invite.py:1]: add created_at and deleted_at fields to dataclass, add __post_init__ for created_at field, add this fields to to_grpc_invite method using FromDatetime. Check event.py code in master branch for instructions

File: invite_service/repository/mock_invite_repository.py [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+
+        Raises
+        ------
+        prisma.errors.PrismaError
+            Catch all for every exception raised by Prisma Client Python.
+        ValueNotFoundError
+            No invites was found for given author id.
+
+        """
+        invites = [invite for invite in self._invites if invite.author_id == author_id]
```
Comment (ZixMai) [inline invite_service/repository/mock_invite_repository.py]: check for deleted_at is None

File: invite_service/repository/mock_invite_repository.py [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+        Raises
+        ------
+        Value Not Found Error
+            Can't delete invite with provided data
+
+        """
+        index = next(
+            i
+            for i in range(len(self._invites))
+            if self._invites[i].id == invite_id
```
Comment (ZixMai) [inline invite_service/repository/mock_invite_repository.py]: check for deleted_at is None

**Review by ZixMai — CHANGES_REQUESTED** [review]
Checked by Elena Letuchaya.

File: invite_service/src/invite_service_impl.py [hunk 12]
```diff
@@ -73,6 +70,32 @@ def get_invites_by_user_id(
             ),
         )
 
+    def get_all_invites(
+        self, request: None, context: grpc.ServicerContext
```
Comment (ZixMai) [inline invite_service/src/invite_service_impl.py]: 0_0 request: None? Make a new message in proto file or use empty type from google

File: invite_service/repository/invite_repository_impl.py [hunk 13]
```diff
@@ -68,15 +68,17 @@ async def get_invites(self, author_id: str) -> List[Invite]:
...
             raise ValueNotFoundError("Invites not found")
         return [
             Invite.from_prisma_invite(prisma_invite=db_invite)
             for db_invite in db_invites
         ]
 
-    async def get_all_invites(self, invite: Invite) -> List[Invite]:
+    async def get_all_invites(self) -> List[Invite]:
         """
         Get all invites.
```
Comment (ZixMai) [inline invite_service/repository/invite_repository_impl.py:84]: remove parameters from docstring...

**Review by ZixMai — CHANGES_REQUESTED** [review]
add post init !!!!! check my comments and fix all problems

File: invite_service/repository/invite_repository_interface.py [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+        pass
+
+    @abstractmethod
+    async def get_all_invites(self) -> List[Invite]:
+        """
+        Get all invites.
+
+        Parameters
+        ----------
+        invite : Invite
```
Comment (ZixMai) [inline invite_service/repository/invite_repository_interface.py]: remove parameters from docstring

File: invite_service/repository/mock_invite_repository.py [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+            return self._invites
+        raise ValueNotFoundError("Invites not found")
+
+    async def get_invites_by_invitee_id(self, invitee_id: List[str]) -> List[Invite]:
+        """
+        Get invites by invitee id.
+
+        Parameters
+        ----------
+        invitee_id : List[str]
```
Comment (ZixMai) [inline invite_service/repository/mock_invite_repository.py]: it takes str, not List[str] change it in the args and in docstring

**Review by ZixMai — CHANGES_REQUESTED** [review]
:(

**Review by ZixMai — APPROVED** [review]

**Review by MaxPlays35 — APPROVED** [review]
Done very well

File: invite_service/repository/invite_repository_impl.py [hunk 16]
```diff
@@ -0,0 +1,199 @@
...
+        ]
+
+    async def get_invites_by_invitee_id(self, invitee_id: str) -> List[Invite]:
+        """
+        Get invites by invitee id.
+
+        Parameters
+        ----------
+        invitee_id : str
+            invitee_id object.
```
Comment (Nhsdkk) [inline invite_service/repository/invite_repository_impl.py]: Should be invitee id

File: invite_service/repository/invite_repository_impl.py [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+            Catch all for every exception raised by Prisma Client Python.
+        ValueNotFoundError
+            No invites was found for given invitee id.
+
+        """
+        db_invites: Optional[
+            List[PrismaInvite]
+        ] = await self._db_client.db.invite.find_many(
+            where={
+                "id": {"in": [invite_id for invite_id in invitee_id]},
```
Comment (Nhsdkk) [inline invite_service/repository/invite_repository_impl.py]: Should be "id": invitee_id,

File: invite_service/repository/invite_repository_interface.py [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+
+    @abstractmethod
+    async def get_invites_by_invitee_id(self, invitee_id: str) -> List[Invite]:
+        """
+        Get invites by invitee id.
+
+        Parameters
+        ----------
+        invitee_id : str
+            invitee_id object.
```
Comment (Nhsdkk) [inline invite_service/repository/invite_repository_interface.py]: Should be just invitee id

File: invite_service/repository/invite_repository_impl.py [hunk 19]
```diff
@@ -0,0 +1,199 @@
...
+    -------
+    async get_invites(author_id)
+        Returns invites that has matches with given author id.
+    async create_invite(invite)
+        Creates new invite inside db or throws an exception.
+    async update_invite(incite)
+        Updates invite that has the same id as provided invite object inside db or throws an exception.
+    async delete_invite(invite_id)
+        Deletes invite that has matching id from database or throws an exception.
+    async get_all_invites()
```
Comment (Nhsdkk) [inline invite_service/repository/invite_repository_impl.py:35]: Sort please

File: invite_service/repository/mock_invite_repository.py [hunk 20]
```diff
@@ -0,0 +1,173 @@
...
+    Mock class for manipulating with invites data
+
+    Attributes
+    ----------
+    _invites: List[Invite]
+        List of invites
+
+    Methods
+    -------
+    async get_invites(author_id)
```
Comment (Nhsdkk) [inline invite_service/repository/mock_invite_repository.py]: rename function to get_invites_by_author_id

File: invite_service/repository/invite_repository_interface.py [hunk 21]
```diff
@@ -0,0 +1,151 @@
...
+from src.models.invite import Invite
+
+
+class InviteRepositoryInterface(ABC):
+    """
+    Interface for class for manipulating with invite data
+
+    Methods
+    -------
+    async get_invites(author_id)
```
Comment (Nhsdkk) [inline invite_service/repository/invite_repository_interface.py]: rename function to get_invites_by_author_id

File: invite_service/repository/invite_repository_interface.py [hunk 22]
```diff
@@ -0,0 +1,151 @@
...
+        Deletes invite that has matching id from database or throws an exception.
+    async get_all_invites()
+        Returns all invites.
+    async get_invites_by_invitee_id(invitee_id)
+        Returns invites that has matches with given invitee id.
+
+    """
+
+    @abstractmethod
+    async def get_invites(self, author_id: str) -> List[Invite]:
```
Comment (Nhsdkk) [inline invite_service/repository/invite_repository_interface.py]: rename function to get_invites_by_author_id

File: invite_service/repository/invite_repository_impl.py [hunk 23]
```diff
@@ -0,0 +1,199 @@
...
+    Class for manipulating with invite data
+
+    Attributes
+    ----------
+    _db_client : prisma.Client
+        Postgres db client.
+
+    Methods
+    -------
+    async get_invites(author_id)
```
Comment (Nhsdkk) [inline invite_service/repository/invite_repository_impl.py]: rename function to get_invites_by_author_id

File: invite_service/repository/invite_repository_impl.py [hunk 24]
```diff
@@ -0,0 +1,199 @@
...
+        Returns invites that has matches with given invitee id.
+
+    """
+
+    _db_client: PostgresClient
+
+    def __init__(self) -> None:
+        self._db_client = PostgresClient()
+
+    async def get_invites(self, author_id: str) -> List[Invite]:
```
Comment (Nhsdkk) [inline invite_service/repository/invite_repository_impl.py]: rename function to get_invites_by_author_id

File: invite_service/repository/mock_invite_repository.py [hunk 25]
```diff
@@ -0,0 +1,173 @@
...
+        Deletes invite that has matching id from database.
+
+    """
+
+    _invites: List[Invite]
+
+    def __init__(self) -> None:
+        self._invites = []
+
+    async def get_invites(self, author_id: str) -> List[Invite]:
```
Comment (Nhsdkk) [inline invite_service/repository/mock_invite_repository.py]: rename function to get_invites_by_author_id

File: invite_service/repository/mock_invite_repository.py [hunk 26]
```diff
@@ -0,0 +1,173 @@
...
+        Raises
+        ------
+        ValueNotFoundError
+            No invites was found for given author id.
+
+        """
+        invites = [
+            invite
+            for invite in self._invites
+            if invite.id in invitee_id and invite.deleted_at is None
```
Comment (Nhsdkk) [inline invite_service/repository/mock_invite_repository.py]: Should be if invite.id == invitee_id and invite.deleted_at is None

File: invite_service/repository/mock_invite_repository.py [hunk 27]
```diff
@@ -0,0 +1,173 @@
...
+
+        Raises
+        ------
+        ValueNotFoundError
+            Can't update invite with provided data
+
+        """
+        index = next(
+            i for i in range(len(self._invites)) if self._invites[i].id == invite.id
+        )
```
Comment (Nhsdkk) [inline invite_service/repository/mock_invite_repository.py]: Catch StopIteration and throw ValueNotFoundError

File: invite_service/repository/mock_invite_repository.py [hunk 28]
```diff
@@ -0,0 +1,173 @@
...
+        if self._invites[index].author_id == invite.author_id:
```
Comment (Nhsdkk) [inline invite_service/repository/mock_invite_repository.py]: Permission check should not be on repository level

File: invite_service/src/invite_service_impl.py [hunk 29]
```diff
@@ -27,20 +27,17 @@ class InviteServiceImpl(GrpcServicer):
...
-            Invite(
-                "id", "event_id", "author_id", "invitee_id", proto.InviteStatus.PENDING
-            ),
-        ]
+    _invite_repository: InviteRepositoryInterface
+
+    def __init__(self, invite_repository: InviteRepositoryInterface) -> None:
+        self._invite_repository = invite_repository
 
     def get_invites_by_user_id(
```
Comment (Nhsdkk) [inline invite_service/src/invite_service_impl.py]: rename function to get_invites_by_author_id

**Review by Nhsdkk — CHANGES_REQUESTED** [review]
Make changes

**Review by Nhsdkk — APPROVED** [review]
LGTM

**Pull request merged**